Intermittent failure in macOS x86-64 test_window_state_change

### Describe the bug

The testbed test `tests/window/test_window.py::test_window_state_change[MainWindow-second_window_kwargs0-WindowState.FULLSCREEN-WindowState.MAXIMIZED]` fails intermittently.

### Steps to reproduce

Run CI; the macOS x86_64 test suite fails intermittently. 3 of 4 dependabot runs failed on 16 Nov. 

### Expected behavior

Window state change tests should pass consistently.

### Screenshots

_No response_

### Environment

- Operating System: macOS (x86_64)
- Python version: 3.12


### Logs

```
_ test_window_state_change[MainWindow-second_window_kwargs0-WindowState.FULLSCREEN-WindowState.MAXIMIZED] _
Traceback (most recent call last):
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/runner.py", line 353, in from_call
    result: TResult | None = func()
                             ^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/runner.py", line 245, in <lambda>
    lambda: runtest_hook(item=item, **kwds),
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/logging.py", line 850, in pytest_runtest_call
    yield
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/capture.py", line 900, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/skipping.py", line 268, in pytest_runtest_call
    return (yield)
            ^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/runner.py", line 179, in pytest_runtest_call
    item.runtest()
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pytest_asyncio/plugin.py", line 469, in runtest
    super().runtest()
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/python.py", line 1720, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_hooks.py", line 512, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 167, in _multicall
    raise exception
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 139, in _multicall
    teardown.throw(exception)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 53, in run_old_style_hookwrapper
    return result.get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_result.py", line 103, in get_result
    raise exc.with_traceback(tb)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 38, in run_old_style_hookwrapper
    res = yield
          ^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pluggy/_callers.py", line 121, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/_pytest/python.py", line 166, in pytest_pyfunc_call
    result = testfunction(**testargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app_packages/pytest_asyncio/plugin.py", line 716, in inner
    runner.run(coro, context=context)
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app/tests/conftest.py", line 167, in run_until_complete
    return asyncio.run_coroutine_threadsafe(coro, self.loop).result()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Frameworks/Python.framework/Versions/3.12/lib/python3.12/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app/tests/window/test_window.py", line 912, in test_window_state_change
    raise closure_exception
  File "/Users/runner/work/toga/toga/testbed/build/testbed/macos/app/Toga Testbed.app/Contents/Resources/app/tests/window/test_window.py", line 831, in check_final_state_size
    assert current_size < previous_state_window_size
AssertionError: assert Size(width=1920.0, height=1055.0) < Size(width=1920.0, height=1028.0)
---------------------------- Captured stdout setup -----------------------------
Closing window
Resetting main_window
Window (Secondary Window) has been created
----------------------------- Captured stdout call -----------------------------
Secondary window is visible
Secondary window is in WindowState.FULLSCREEN
Secondary window is in WindowState.MAXIMIZED
=========================== short test summary info ============================
FAILED tests/window/test_window.py::test_window_state_change[MainWindow-second_window_kwargs0-WindowState.FULLSCREEN-WindowState.MAXIMIZED] - assert Size(width=1920.0, height=1055.0) < Size(width=1920.0, height=1028.0)
= 1 failed, 602 passed, 4 skipped, 37 xfailed, 2 xpassed in 411.51s (0:06:51) ==
```


### Additional context

The failure suggests the "maximized" state hasn't fully taken effect yet; a longer delay might correct the problem.

--- a/changes/3897.misc.md
+++ b/changes/3897.misc.md
@@ -1,0 +1,1 @@
+The Cocoa testbed's window state probe now waits for a window's reported size to stabilize (not just its state flag) before considering a state transition complete, fixing an intermittent failure in `test_window_state_change` on macOS x86-64 CI runners.

--- a/cocoa/tests_backend/window.py
+++ b/cocoa/tests_backend/window.py
@@ -39,19 +39,42 @@ class WindowProbe(BaseProbe, DialogsMixin):
         if state:
             timeout = 5
             polling_interval = 0.1
+            # Number of consecutive polls the window's size must be unchanged for
+            # before considering the state transition fully settled.
+            required_stable_polls = 2
             exception = None
+            stable_polls = 0
+            last_size = None
             loop = asyncio.get_running_loop()
             start_time = loop.time()
             while (loop.time() - start_time) < timeout:
                 try:
                     assert self.instantaneous_state == state
                     assert self.window._impl._pending_state_transition is None
-                    return
+                    # AppKit can report a window as having completed a state
+                    # transition (e.g. via windowDidEnterFullScreen:) slightly
+                    # before its on-screen frame has finished animating to the
+                    # final size, especially on slower CI runners. Waiting for
+                    # the reported size to be stable across a couple of polls
+                    # avoids capturing a transient, not-yet-settled size (#3897).
+                    current_size = self.window.size
+                    if current_size == last_size:
+                        stable_polls += 1
+                        if stable_polls >= required_stable_polls:
+                            return
+                    else:
+                        stable_polls = 0
+                    last_size = current_size
                 except AssertionError as e:
                     exception = e
-                    await asyncio.sleep(polling_interval)
-                    continue
+                    stable_polls = 0
+                    last_size = None
+                await asyncio.sleep(polling_interval)
+            if exception:
                 raise exception
+            raise AssertionError(
+                f"Window size did not stabilize in {state} state within {timeout}s"
+            )
 
     async def cleanup(self):
         # Store the pre closing window state as determination of window state after

--- a/cocoa/tests_backend/window.py
+++ b/cocoa/tests_backend/window.py
@@ -40,7 +40,7 @@ class WindowProbe(BaseProbe, DialogsMixin):
             timeout = 5
             polling_interval = 0.1
             # Number of consecutive polls the window's size must be unchanged for
-            # before considering the state transition fully settled.
+            # before considering the state transition fully settled. See #3897.
             required_stable_polls = 2
             exception = None
             stable_polls = 0
@@ -51,12 +51,6 @@ class WindowProbe(BaseProbe, DialogsMixin):
                 try:
                     assert self.instantaneous_state == state
                     assert self.window._impl._pending_state_transition is None
-                    # AppKit can report a window as having completed a state
-                    # transition (e.g. via windowDidEnterFullScreen:) slightly
-                    # before its on-screen frame has finished animating to the
-                    # final size, especially on slower CI runners. Waiting for
-                    # the reported size to be stable across a couple of polls
-                    # avoids capturing a transient, not-yet-settled size (#3897).
                     current_size = self.window.size
                     if current_size == last_size:
                         stable_polls += 1

